GVFS/GVFS/CommandLine/SparseVerb.cs

```diff
@@ -235,10 +237,105 @@ protected override void Execute(GVFSEnlistment enlistment)
                     {
                         this.PruneFoldersOutsideSparse(tracer, enlistment, sparseTable);
                     }
+
+                    if (needToChangeProjection || this.Prune)
+                    {
+                        // Update the last write times of the parents of folders being added/removed
+                        // so that File Explorer will refresh them
+                        UpdateParentFolderLastWriteTimes(tracer, enlistment.WorkingDirectoryBackingRoot, foldersToRemove, foldersToAdd);
+                    }
                 }
             }
         }
 
+        private static void UpdateParentFolderLastWriteTimes(ITracer tracer, string rootPath, List<string> foldersToRemove, List<string> foldersToAdd)
+        {
+            Stopwatch updateTime = Stopwatch.StartNew();
+
+            HashSet<string> foldersToRefresh = new HashSet<string>(GVFSPlatform.Instance.Constants.PathComparer);
+            foreach (string path in foldersToRemove)
+            {
+                AddNonRootParentPathsToSet(foldersToRefresh, path);
+            }
+
+            foreach (string path in foldersToAdd)
+            {
+                AddNonRootParentPathsToSet(foldersToRefresh, path);
+            }
+
+            DateTime refreshTime = DateTime.Now;
+            int foldersRefreshed = 0;
+            int refreshErrors = 0;
+            bool refreshFailed;
+
+            // Always refresh the root
+            if (SetFolderLastWriteTimeIfOnDisk(tracer, rootPath, refreshTime, out refreshFailed))
+            {
+                ++foldersRefreshed;
+            }
+            else if (refreshFailed)
+            {
+                ++refreshErrors;
+            }
+
+            string folderPathPrefix = $"{rootPath}{Path.DirectorySeparatorChar}";
+            foreach (string path in foldersToRefresh)
+            {
+                if (SetFolderLastWriteTimeIfOnDisk(tracer, folderPathPrefix + path, refreshTime, out refreshFailed))

```

Comment:
I had to go back up to the code to figure out why this is returning a `bool` and `refreshFailed` and what the difference between returning `false` and `refreshFailed` being `false` so maybe a comment would help.

Reply:
@kewillford two other alternatives I considered for this method:

-  Make the method `void` and have an `out` parameter `directoryExists` (in addition to `refreshFailed`)
- Have the method return an result enum with possible values `Success`, `Failure`, `DirectoryDoesNotExist`

Do you think either of those would be preferable to the current approach?  Initially there was no `refreshFailed`, but I decided that it would be nice to log failures specifically (and not lump their count together with the directory not existing).

Reply:
Yeah we really have 3 different results here.
1. The directory was refreshed
2. The directory didn't exist
3. Tried to refresh the directory but it failed

I'm kind of leaning toward the result enum because it gives us the 3 states vs the 2 bools which give us 4 one of which is never possible.

Do we want to log the directories that don't exist?

Reply:
> Do we want to log the directories that don't exist?

It will be relatively common for some of the directories to not be on disk, and so I think we're ok just logging the counts.  It would be tough to differentiate between when it's expected and unexpected that a directory be on disk.